workflow/gccontroller/gc_controller.go

```diff
@@ -0,0 +1,336 @@
+package gccontroller
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/clock"
+	runtimeutil "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/argoproj/argo-workflows/v3/config"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	wfclientset "github.com/argoproj/argo-workflows/v3/pkg/client/clientset/versioned"
+	commonutil "github.com/argoproj/argo-workflows/v3/util"
+	envutil "github.com/argoproj/argo-workflows/v3/util/env"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/metrics"
+	"github.com/argoproj/argo-workflows/v3/workflow/util"
+)
+
+type Controller struct {
+	wfclientset           wfclientset.Interface
+	wfInformer            cache.SharedIndexInformer
+	workqueue             workqueue.DelayingInterface
+	clock                 clock.Clock
+	metrics               *metrics.Metrics
+	completedOrderedQueue *priorityQueue
+	failedOrderedQueue    *priorityQueue
+	errorOrderedQueue     *priorityQueue
+	retentionPolicy       *config.RetentionPolicy
+	retentionGCPeriod     time.Duration
+}
+
+type priorityQueue struct {
+	sync.Mutex
+	heap  []*unstructured.Unstructured
```

Comment:
Based upon the deduplication logic I think we should keep it as unstructured since that is what we expect it to be: https://github.com/argoproj/argo-workflows/blob/018a64d6491fe16a1c56c14c4b4309ed58f45d70/workflow/gccontroller/heap.go#L29-L33